lib/rules/no-inner-declarations.js

```diff
@@ -34,53 +37,48 @@ module.exports = {
     create(context) {
 
         /**
-         * Find the nearest Program or Function ancestor node.
-         * @returns {Object} Ancestor's type and distance from node.
+         * Ensure that a given node is at a program or function body's root.
+         * @param {ASTNode} node Declaration node to check.
+         * @returns {void}
          */
-        function nearestBody() {
+        function check(node) {
             const ancestors = context.getAncestors();
-            let ancestor = ancestors.pop(),
-                generation = 1;
+            let ancestor = ancestors.pop();
 
-            while (ancestor && ["Program", "FunctionDeclaration",
-                "FunctionExpression", "ArrowFunctionExpression"
-            ].indexOf(ancestor.type) < 0) {
-                generation += 1;
-                ancestor = ancestors.pop();
+            if (
+                ancestor.type === "BlockStatement" &&
+            !!~validBlockStatementParent.indexOf(ancestor.parent.type)
```

Comment:
Can we use `validBlockStatementParent.includes` here, I think it's more readable than `!!~validBlockStatementParent.indexOf`.

Alternatively, `validBlockStatementParent` might be a `Set` instead of an array.

(the same applies to the other condition below)

Reply:
> Alternatively, `validBlockStatementParent` might be a `Set` instead of an array.

This seems much better 👍 I will change it soon

Reply:
done